packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -74,35 +75,47 @@ library LibProposing {
             revert L1_TOO_MANY_BLOCKS();
         }
 
-        bytes32 parentMetaHash =
-            _state.blocks[(b.numBlocks - 1) % _config.blockRingBufferSize].metaHash;
-        // assert(parentMetaHash != 0);
+        TaikoData.Block storage parentBlock =
+            _state.blocks[(b.numBlocks - 1) % _config.blockRingBufferSize];
 
         // Check if parent block has the right meta hash. This is to allow the proposer to make sure
         // the block builds on the expected latest chain state.
-        if (params.parentMetaHash != 0 && parentMetaHash != params.parentMetaHash) {
+        if (params.parentMetaHash != 0 && parentBlock.metaHash != params.parentMetaHash) {
             revert L1_UNEXPECTED_PARENT();
         }
 
+        // Verify the passed in L1 state block number.
+        // We only allow the L1 block to be 2 epochs old.
```

Comment:
is "2 epoch" a special value? 

If we use epoch, shall we check `params.l1StateBlockNumber/32 >= block.number/32 - 2`

Reply:
No real special value, but this window should be large enough for a preconfer to always be able to select an L1 block that he can later propose in his own slot. That should only be up to 2 epochs of L1 slots.